nexus_system_orchestrator/src/context.hpp

```diff
@@ -29,47 +29,267 @@
 #include <rclcpp_action/rclcpp_action.hpp>
 #include <rclcpp_lifecycle/lifecycle_node.hpp>
 
+#include <mutex>
+#include <optional>
+
 namespace nexus::system_orchestrator {
 
 class Context
 {
-public: using WorkOrder = common::WorkOrder;
-public: using WorkOrderGoalHandle =
+public:
+  using WorkOrder = common::WorkOrder;
+  using WorkOrderGoalHandle =
     rclcpp_action::ServerGoalHandle<nexus::endpoints::WorkOrderAction::ActionType>;
-public: using WorkcellTask = nexus_orchestrator_msgs::msg::WorkcellTask;
-public: using TaskState = nexus_orchestrator_msgs::msg::TaskState;
+  using WorkcellTask = nexus_orchestrator_msgs::msg::WorkcellTask;
+  using TaskState = nexus_orchestrator_msgs::msg::TaskState;
+
+  Context(rclcpp_lifecycle::LifecycleNode& node) : _node(node) {}
+
+  rclcpp_lifecycle::LifecycleNode& get_node() const
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    return _node;
+  }
+
+  Context& set_job_id(const std::string& job_id)
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    _job_id = job_id;
+    return *this;
+  }
+
+  std::string get_job_id() const
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    return _job_id;
+  }
+
+  Context& set_work_order(const WorkOrder& wo)
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    _wo = wo;
+    return *this;
+  }
+
+  WorkOrder get_work_order() const
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    return _wo;
+  }
+
+  Context& set_tasks(const std::vector<WorkcellTask>& tasks)
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    _tasks = tasks;
+    return *this;
+  }
+
+  std::vector<WorkcellTask> get_tasks() const
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    return _tasks;
+  }
+
+  std::size_t get_task_number() const
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    return _tasks.size();
+  }
+
+  Context& set_task_remapper(const std::shared_ptr<const common::TaskRemapper>& remapper)
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    _task_remapper = remapper;
+    return *this;
+  }
+
+  std::shared_ptr<const common::TaskRemapper> get_task_remapper() const
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    return _task_remapper;
+  }
+
+  Context& set_workcell_task_assignment(const std::string& task_id, const std::string& workcell_id)
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    _workcell_task_assignments[task_id] = workcell_id;
```

Comment:
It might be good to check if `task_id` already has an assigned workcell first and log a warning if so before overwriting.

```suggestion
    if (_workcell_task_assignments.count(task_id) > 0)
    {
      RCLCPP_WARN(
        _node.get_logger(),
        [Context::set_workcell_task_assignment] task_id %s previously assigned to %s is now reassigned to %s.,
        task_id.c_str(), _workcell_task_assignments[task_id].c_str(), workcell_id.c_str()
      );
    }
    _workcell_task_assignments[task_id] = workcell_id;
```

Reply:
That makes sense, https://github.com/osrf/nexus/pull/75/commits/1a073a5645b7fd2b0298ec130eb3845ed805dece